File: packages/core/src/middleware/session-echo.ts
import { SessionManager } from "../session.js";

export function wrapResponse(result: any, params: any, toolName: string, sessionManager: SessionManager) {
    const sessionId = params.session_id;
    if (!sessionId) {
        // If it's a 'begin' action, it returns session_id in the result
        if (toolName === "pg_tx" && params.action === "begin" && result.session_id) {
            return {
                ...result,
                active_session: {
                    id: result.session_id,
                    hint: "Use this session_id for subsequent transactional queries."
                }
            };
        }
        return result;
    }

    const sessions = sessionManager.listSessions();
    const currentSession = sessions.find(s => s.id === sessionId);
    
    if (!currentSession) return result;

    // Check if we should echo based on the plan:
    // 1. Write operations (action: 'write', 'ddl', 'set', etc.)
    // 2. Near expiry (< 5 minutes)
    
    const isWriteOp = 
        (toolName === "pg_query" && (params.action === "write" || params.action === "transaction")) ||
        (toolName === "pg_schema" && (params.action === "create" || params.action === "alter" || params.action === "drop")) ||
        (toolName === "pg_admin" && params.subaction === "set");

    const expiresInMinutes = parseInt(currentSession.expires_in);
    const isNearExpiry = expiresInMinutes < 5;

    if (isWriteOp || isNearExpiry) {
        return {
            ...result,
            active_session: {
                id: currentSession.id,
                expires_in: currentSession.expires_in,
                hint: isNearExpiry 
                    ? "Warning: Session expiring soon. Commit your work shortly." 
                    : `Active transaction: ${currentSession.id}`
            }
        };
    }

    return result;
}

File: coldquery/middleware/session_echo.py
import json
from typing import Any, Dict, Optional

from coldquery.core.session import SessionManager


def enrich_response(
    result: Dict[str, Any], session_id: Optional[str], session_manager: SessionManager
) -> str:
    """
    Enriches the response with session metadata.

    Args:
        result: The original result dictionary.
        session_id: The session ID for the current operation.
        session_manager: The session manager instance.

    Returns:
        A JSON string representing the enriched response.
    """
    if not session_id:
        return json.dumps(result)

    session = session_manager.get_session(session_id)
    if not session:
        return json.dumps(result)

    expires_in_minutes = session.expires_in
    is_near_expiry = expires_in_minutes < 5

    if is_near_expiry:
        result["active_session"] = {
            "id": session.id,
            "expires_in": f"{expires_in_minutes}m",
            "hint": "Warning: Session expiring soon. Commit your work shortly.",
        }

    return json.dumps(result)
